pkg/actuators/awsplacementgroup/controller.go

```diff
@@ -3,91 +3,623 @@ package awsplacementgroup
+	clusterID := infra.Status.InfrastructureName
+
+	for _, tag := range mergedTags {
+		// AWS tags are case sensitive, so we don't need to worry about other casing of "Name"
+		if !strings.HasPrefix(tag.Name, "kubernetes.io/cluster/") && tag.Name != "Name" {
+			rawTagList = append(rawTagList, &ec2.Tag{Key: aws.String(tag.Name), Value: aws.String(tag.Value)})
+		}
+	}
+
+	rawTagList = append(rawTagList, []*ec2.Tag{
+		{Key: aws.String("kubernetes.io/cluster/" + clusterID), Value: aws.String("owned")},
+		{Key: aws.String("Name"), Value: aws.String(awsPlacementGroup)},
+	}...)
+
+	return removeDuplicatedTags(rawTagList)
+}
+
+// removeDuplicatedTags scan machine tags, and return a deduped tags list. The first found value gets precedence.
+func removeDuplicatedTags(tags []*ec2.Tag) []*ec2.Tag {
+	m := make(map[string]bool)
+	result := []*ec2.Tag{}
+
+	// look for duplicates
+	for _, entry := range tags {
+		if _, value := m[*entry.Key]; !value {
+			m[*entry.Key] = true
+
+			result = append(result, entry)
+		}
+	}
+
+	return result
+}
+
+// isAWS4xxError will determine if the passed error is an AWS error with a 4xx status code.
+func isAWS4xxError(err error) bool {
+	if _, ok := err.(awserr.Error); ok {
+		if reqErr, ok := err.(awserr.RequestFailure); ok {
+			if reqErr.StatusCode() >= 400 && reqErr.StatusCode() < 500 {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// checkOrCreatePlacementGroup checks for the existence of a placement group on AWS and validates its config
+// it proceeds to create one if such group doesn't exist.
+func checkOrCreatePlacementGroup(client awsclient.Client, logger logr.Logger, pg *machinev1.AWSPlacementGroup, infra *configv1.Infrastructure) error {
+	placementGroups, err := client.DescribePlacementGroups(&ec2.DescribePlacementGroupsInput{
+		GroupNames: []*string{aws.String(pg.Name)},
+	})
+	if err != nil && !isAWS4xxError(err) {
+		// Ignore a 400 error as AWS will report an unknown placement group as a 400.
+		return fmt.Errorf("failed to check aws placement group: could not describe aws placement groups: %w", err)
+	}
+
+	// More than one placement group matching.
+	if len(placementGroups.PlacementGroups) > 1 {
+		return fmt.Errorf("failed to check aws placement group: expected 1 aws placement group for name %q, got %d", pg.Name, len(placementGroups.PlacementGroups))
+	}
+
+	// Placement group already exists on AWS.
+	if len(placementGroups.PlacementGroups) == 1 {
+		// Validate it's configuration.
+		if err := validateExistingPlacementGroupConfig(pg, placementGroups.PlacementGroups[0]); err != nil {
+			// Set the Ready Condition to False due to a Configuration Mismatch.
+			meta.SetStatusCondition(&pg.Status.Conditions, metav1.Condition{
+				Type:    readyConditionType,
+				Status:  metav1.ConditionFalse,
+				Reason:  configurationMismatchConditionReason,
+				Message: err.Error(),
+			})
+
+			return err
```

Comment:
Add more context to this error please

Reply:
Done. I refactored a bit `validateExistingPlacementGroupConfig()`'s inner error messages to read better with the newly added outer error context. See: https://github.com/openshift/machine-api-provider-aws/pull/33/commits/a9c9cdc2c824c33d6ce6f5c48e9509fb26c168d6